components/feature/media/src/main/java/mozilla/components/feature/media/MediaSessionFeature.kt

```diff
@@ -0,0 +1,92 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.media
+
+import android.app.Activity
+import android.content.Context
+import android.content.pm.ActivityInfo
+import androidx.core.content.ContextCompat
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.map
+import mozilla.components.browser.state.state.SessionState
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.feature.media.service.AbstractMediaSessionService
+import mozilla.components.lib.state.ext.flowScoped
+import mozilla.components.support.base.feature.LifecycleAwareFeature
+import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
+
+/**
+ * Feature implementation that handles MediaSession state changes
+ *
+ * @param activity the application's [Activity].
+ * @param mediaServiceClass the media service class will handle the media playback state
+ * @param store Reference to the browser store where tab state is located.
+ * @param applicationContext the application's [Context].
+ */
+class MediaSessionFeature(
+    val activity: Activity,
+    val mediaServiceClass: Class<*>,
+    val store: BrowserStore,
+    val applicationContext: Context = activity.applicationContext
+
+) : LifecycleAwareFeature {
```

Comment:
You are mixing two things with different lifecycles here:
* Activity scope: Fullscreen mode depends on the activity and is only relevant for the current tab while you look at the tab.
* Global scope: Starting the service is a global thing. It needs to happen regardless of whether an activity is being displayed (e.g. app in the background starts playing something) and it would be wasteful to do it multiple times if there would be multiple activities/features active.

I think you want to split this.

Reply:
After investigation, you're right, I'll split the file up.